core/src/main/java/bisq/core/trade/protocol/bsqswap/tasks/buyer/BuyerFinalizeTx.java

```diff
@@ -0,0 +1,151 @@
+import bisq.core.trade.protocol.bsqswap.BsqSwapTradePeer;
+import bisq.core.trade.protocol.bsqswap.tasks.BsqSwapTask;
+
+import bisq.common.taskrunner.TaskRunner;
+
+import org.bitcoinj.core.Address;
+import org.bitcoinj.core.Coin;
+import org.bitcoinj.core.NetworkParameters;
+import org.bitcoinj.core.Transaction;
+import org.bitcoinj.core.TransactionInput;
+import org.bitcoinj.core.TransactionOutPoint;
+import org.bitcoinj.core.TransactionOutput;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+import lombok.extern.slf4j.Slf4j;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+@Slf4j
+public class BuyerFinalizeTx extends BsqSwapTask {
+    @SuppressWarnings({"unused"})
+    public BuyerFinalizeTx(TaskRunner<BsqSwapTrade> taskHandler, BsqSwapTrade bsqSwapTrade) {
+        super(taskHandler, bsqSwapTrade);
+    }
+
+    @Override
+    protected void run() {
+        try {
+            runInterceptHook();
+
+            BsqSwapTradePeer tradePeer = bsqSwapProtocolModel.getTradePeer();
+            TradeWalletService tradeWalletService = bsqSwapProtocolModel.getTradeWalletService();
+            BtcWalletService btcWalletService = bsqSwapProtocolModel.getBtcWalletService();
+            NetworkParameters params = btcWalletService.getParams();
+
+            Transaction sellersTransaction = btcWalletService.getTxFromSerializedTx(tradePeer.getTx());
+
+            // The inputs from the deserialized tx do not have connected outpoints and no value assigned.
+            // We reapply that by using the RawTransactionInputs
+            int buyersInputSize = Objects.requireNonNull(bsqSwapProtocolModel.getInputs()).size();
+            List<TransactionInput> sellersInputs = sellersTransaction.getInputs().stream()
+                    .filter(i -> i.getIndex() >= buyersInputSize)
+                    .collect(Collectors.toList());
+
+            List<RawTransactionInput> rawInputs = Objects.requireNonNull(tradePeer.getInputs());
+            List<TransactionInput> sellersBtcInputs = new ArrayList<>();
+            for (int i = 0; i < sellersInputs.size(); i++) {
+                TransactionInput transactionInput = sellersInputs.get(i);
+                RawTransactionInput rawTransactionInput = rawInputs.get(i);
+                Coin value = Coin.valueOf(rawTransactionInput.value);
+                Transaction parentTx = new Transaction(params, rawTransactionInput.parentTransaction);
+                TransactionOutPoint transactionOutPoint = new TransactionOutPoint(params, rawTransactionInput.index, parentTx);
+                TransactionInput result = new TransactionInput(params,
+                        parentTx,
+                        transactionInput.getScriptBytes(),
+                        transactionOutPoint,
+                        value);
+                result.setWitness(transactionInput.getWitness());
```

Comment:
Is this ok for non segwit inputs? It looks fine from what I understand but good to make sure.

Reply:
Yes I will add a ` if (transactionInput.hasWitness())`. I guess its just null otherwise, but better to make it more explicit. We set the ScriptBytes before so those are null in segwit case but otherwise the sig.

Btw. I think we might have an issue as we expect the txId to be independent of sigs, but with non-segwit thats not the case... I guess its only at the tx listener code, which should be enhanced anyway by an additional message with the final tx sent to the seller.